In [ ]:
# import os
import re
import sys
import pandas as pd
import sentencepiece as spm
from read_file import ReadFile

In [3]:
# os.remove('/kaggle/working/english_corpus.txt')

In [4]:
reader = ReadFile(dropna=True, drop_duplicates=False, random_state=42)

In [ ]:
opus_file = 'data/train/opus_yo_en.tsv'        
raw_file = 'data//train/raw_yo_en.tsv'                
JW300 = 'data/train/JW300_en_yo.csv'                 
bible = 'data/train/bible_yo_en.csv'
org_set = 'data/train/Train.csv'                  
train = 'data/dev/train.tsv'


filepath=[JW300, bible, org_set, opus_file, raw_file, train]

df = reader.read_file(filepath=filepath, delimiter=',')

In [6]:
df.to_csv('yo_en_corpus.txt', index=False, header=False, sep='\n')

In [7]:
with open('/kaggle/working/yo_en_corpus.txt', mode='r') as f:
    yo_en_txt = f.read()

len(yo_en_txt)

122456461

In [8]:
# Tokenize and find unique words for each column
unique_english_words = set(" ".join(df['English']).split())
unique_yoruba_words = set(" ".join(df['Yoruba']).split())

# Calculate the sum of unique words
sum_english_unique_words = len(unique_english_words)
sum_yoruba_unique_words = len(unique_yoruba_words)

# Find the real total (union of both sets)
union = len(unique_english_words.union(unique_yoruba_words))
intersection = len(unique_english_words.intersection(unique_yoruba_words))

print(f"Total unique English words: {sum_english_unique_words - intersection}")
print(f"Total unique Yoruba words: {sum_yoruba_unique_words - intersection}")
print(f"Total unique words: {union - intersection}")

Total unique English words: 224841
Total unique Yoruba words: 192436
Total unique words: 417277


## Withouth new dataset

Total unique English words: 153681

Total unique Yoruba words: 101889

Total unique words: 255570


## With new dataset

Total unique English words: 224841

Total unique Yoruba words: 192436

Total unique words: 417277

In [9]:
input_file = '/kaggle/working/yo_en_corpus.txt'
model_prefix = '/kaggle/working/yo_en_bpe'


spm.SentencePieceTrainer.train(
    input=input_file,
    model_prefix=model_prefix,
    vocab_size=32_000,
    character_coverage=1.0,
    max_sentence_length=999999,
    model_type='bpe',
    pad_id=0,
    unk_id=1,
    bos_id=2,
    eos_id=3,
    pad_piece='<PAD>',
    unk_piece='<UNK>',
    bos_piece='<BOS>',
    eos_piece='<EOS>',
    add_dummy_prefix=True
)

In [10]:
# Load your trained tokenizer model
tokenizer = spm.SentencePieceProcessor(model_file="/kaggle/working/yo_en_bpe.model")

# Test with a sample text
sample_text = """Mr. Bamgbose who stressed that as a member of the House, he has been able to also affect his constituents positively though various empowerment programmes, ranging from agriculture, crafts and many artisans within Badagry Federal Constituency. Mr. Bamgbose thanked Badagrians for giving him the opportunity to serve and promised to continue to deliver more dividends to his people.\nOlùdíje fún ipò asòfin àgbà nínú ẹgbẹ́, All Progressives Congress (APC),asòfin  Smart Adeyemi,ni ó jáwé olúborí nínú ètò ìdìbò asòfin tó wáyé ní apá ilà Gúsù ní ìpínlẹ̀ Kogi.\nThe candidate of the All Progressives Congress (APC), Senator Smart Adeyemi, has won the Kogi West senatorial district election.\nKùkùté kan kì í fọ́ni lépo lẹ́ẹ̀mejì.\nNo one stump can break one's oil-pot twice.
Tall, tall, talling, talking, taller, tallest, speak, speaker, speaking
Èèkàn ẹgbẹ́ Afenifere ilẹ̀ Yoruba, Pa Ayo Adebanjo ti jáde láyé.

Ní òwúrọ̀ ọjọ́, Ẹtì, ọjọ́ Kẹrìnlá, ọdún 2024 ni Adebanjo dágbére fáyé lẹ́ni ọdún mẹ́rìndínlọ́gọ́rùn-ún (96).

Agbẹnusọ ẹgbẹ́ Afenifere, Jare Ajayi ló fìdí ìròyìn náà múlẹ̀ fún BBC News Yorùbá.

Ilé rẹ̀ tó wà ní agbègbè Lekki, ìpínlẹ̀ Eko ni Adebanjo dákẹ́ sí.

Pa Adebanjo bẹ̀rẹ̀ sí ní darí ẹgbẹ́ Afenifere nígbà tí ààrẹ ẹgbẹ́ náà, Bàbá Fasoranti ní kí ó máa delé fún òun látàrí ọjọ́ orí òun.
"""
tokens = tokenizer.encode(sample_text, out_type=str)  # Get tokenized output (as subwords)
print("Tokens:", tokens)

# Decode the tokens back to text
decoded_text = tokenizer.decode(tokens, out_type=str)
print("Decoded Text:", decoded_text)

Tokens: ['▁Mr', '.', '▁B', 'am', 'gb', 'ose', '▁who', '▁stressed', '▁that', '▁as', '▁a', '▁member', '▁of', '▁the', '▁House', ',', '▁he', '▁has', '▁been', '▁able', '▁to', '▁also', '▁affect', '▁his', '▁constit', 'u', 'ents', '▁positively', '▁though', '▁various', '▁empower', 'ment', '▁program', 'mes', ',', '▁ranging', '▁from', '▁agriculture', ',', '▁crafts', '▁and', '▁many', '▁artis', 'ans', '▁within', '▁Bad', 'ag', 'ry', '▁Federal', '▁Cons', 'tit', 'u', 'ency', '.', '▁Mr', '.', '▁B', 'am', 'gb', 'ose', '▁thanked', '▁Bad', 'ag', 'rians', '▁for', '▁giving', '▁him', '▁the', '▁opportunity', '▁to', '▁serve', '▁and', '▁promised', '▁to', '▁continue', '▁to', '▁deliver', '▁more', '▁div', 'id', 'ends', '▁to', '▁his', '▁people', '.', '▁Olù', 'dí', 'je', '▁fún', '▁ipò', '▁asòfin', '▁àgbà', '▁nínú', '▁ẹgbẹ́', ',', '▁All', '▁Progress', 'ives', '▁Congress', '▁(', 'AP', 'C', '),', 'as', 'òfin', '▁Smart', '▁Ade', 'y', 'emi', ',', 'ni', '▁ó', '▁já', 'wé', '▁olúborí', '▁nínú', '▁ètò', '▁ìdìbò', '▁asòfin', 

In [11]:
# Load your trained tokenizer model
tokenizer = spm.SentencePieceProcessor(model_file="/kaggle/working/yo_en_bpe.model")

# Test with a sample text
sample_text = """Mr. Bamgbose who stressed that as a member of the House, he has been able to also affect his constituents positively though various empowerment programmes, ranging from agriculture, crafts and many artisans within Badagry Federal Constituency. Mr. Bamgbose thanked Badagrians for giving him the opportunity to serve and promised to continue to deliver more dividends to his people.\nOlùdíje fún ipò asòfin àgbà nínú ẹgbẹ́, All Progressives Congress (APC),asòfin  Smart Adeyemi,ni ó jáwé olúborí nínú ètò ìdìbò asòfin tó wáyé ní apá ilà Gúsù ní ìpínlẹ̀ Kogi.\nThe candidate of the All Progressives Congress (APC), Senator Smart Adeyemi, has won the Kogi West senatorial district election.\nKùkùté kan kì í fọ́ni lépo lẹ́ẹ̀mejì.\nNo one stump can break one's oil-pot twice.
Tall, tall, talling, talking, taller, tallest, speak, speaker, speaking
Èèkàn ẹgbẹ́ Afenifere ilẹ̀ Yoruba, Pa Ayo Adebanjo ti jáde láyé.

Ní òwúrọ̀ ọjọ́, Ẹtì, ọjọ́ Kẹrìnlá, ọdún 2024 ni Adebanjo dágbére fáyé lẹ́ni ọdún mẹ́rìndínlọ́gọ́rùn-ún (96).

Agbẹnusọ ẹgbẹ́ Afenifere, Jare Ajayi ló fìdí ìròyìn náà múlẹ̀ fún BBC News Yorùbá.

Ilé rẹ̀ tó wà ní agbègbè Lekki, ìpínlẹ̀ Eko ni Adebanjo dákẹ́ sí.

Pa Adebanjo bẹ̀rẹ̀ sí ní darí ẹgbẹ́ Afenifere nígbà tí ààrẹ ẹgbẹ́ náà, Bàbá Fasoranti ní kí ó máa delé fún òun látàrí ọjọ́ orí òun.
"""
tokens = tokenizer.encode(sample_text, out_type=str)  # Get tokenized output (as subwords)
print("Tokens:", tokens)

# Decode the tokens back to text
decoded_text = tokenizer.decode(tokens, out_type=str)
print("Decoded Text:", decoded_text)

Tokens: ['▁Mr', '.', '▁B', 'am', 'gb', 'ose', '▁who', '▁stressed', '▁that', '▁as', '▁a', '▁member', '▁of', '▁the', '▁House', ',', '▁he', '▁has', '▁been', '▁able', '▁to', '▁also', '▁affect', '▁his', '▁constit', 'u', 'ents', '▁positively', '▁though', '▁various', '▁empower', 'ment', '▁program', 'mes', ',', '▁ranging', '▁from', '▁agriculture', ',', '▁crafts', '▁and', '▁many', '▁artis', 'ans', '▁within', '▁Bad', 'ag', 'ry', '▁Federal', '▁Cons', 'tit', 'u', 'ency', '.', '▁Mr', '.', '▁B', 'am', 'gb', 'ose', '▁thanked', '▁Bad', 'ag', 'rians', '▁for', '▁giving', '▁him', '▁the', '▁opportunity', '▁to', '▁serve', '▁and', '▁promised', '▁to', '▁continue', '▁to', '▁deliver', '▁more', '▁div', 'id', 'ends', '▁to', '▁his', '▁people', '.', '▁Olù', 'dí', 'je', '▁fún', '▁ipò', '▁asòfin', '▁àgbà', '▁nínú', '▁ẹgbẹ́', ',', '▁All', '▁Progress', 'ives', '▁Congress', '▁(', 'AP', 'C', '),', 'as', 'òfin', '▁Smart', '▁Ade', 'y', 'emi', ',', 'ni', '▁ó', '▁já', 'wé', '▁olúborí', '▁nínú', '▁ètò', '▁ìdìbò', '▁asòfin', 

In [12]:
print(tokenizer.piece_to_id('<PAD>')) 
print(tokenizer.piece_to_id('<UNK>'))
print(tokenizer.piece_to_id('<EOS>'))
print(tokenizer.piece_to_id('<BOS>'))
print(tokenizer.piece_to_id('</s>'))
print(tokenizer.piece_to_id('#'))

0
1
3
2
1
30356
